### PR TITLE
Fixed #28598: Document lack of BCC in console & filebased email backends

### PR DESCRIPTION
**Pull request by jschneier**
https://code.djangoproject.com/ticket/28598

### CODE REVIEW

**Comment by jschneier** [comment]
@carltongibson do you receive notice when I reply there, I added a few questions/my thoughts on the ticket. Sorry if you do, don't mean to spam.

**Comment by jschneier** [comment]
Is there a policy about whether or not the `..method` doc bits includes `self`? Seems to go with mostly no but occasionally yes.

File: django/core/mail/backends/filebased.py [hunk 1]
```diff
@@ -44,10 +44,13 @@ def __init__(self, *args, file_path=None, **kwargs):
         super().__init__(*args, **kwargs)
 
     def write_message(self, message):
-        self.stream.write(message.message().as_bytes() + b'\n')
+        self.stream.write(b'%s\n' % self.format_message(message))
         self.stream.write(b'-' * 79)
         self.stream.write(b'\n')
 
+    def format_message(self, message):
```
Comment (carltongibson) [inline django/core/mail/backends/filebased.py]: I think `format_message()` should always return a `str`. `write_message()` can `encode()` that if need be. 

Q: Why exactly are we writing bytes here? (NTS: see what fails if you can change this. 🙂)

File: docs/topics/email.txt [hunk 2]
```diff
@@ -529,6 +538,16 @@ To specify this backend, put the following in your settings::
 This backend is not intended for use in production -- it is provided as a
 convenience that can be used during development.
 
+..method:: format_message(message)
```
Comment (carltongibson) [inline docs/topics/email.txt]: Wondering if we can just adjust the doc above to say it subclasses console backend. (?)

**Comment by jschneier** [comment]
@carltongibson here's a different possible iteration since we are in agreement about displaying the BCC. Obviously there might be a backward compatibility issue here. I'm not really in favor of adding the hook you mentioned although I could compromise to that if it's what we need to fix-up backwards compatibility.

Thoughts?

**Comment by jschneier** [comment]
Also curious if your investigation fo the string/bytes has turned up anything? That is the reason that there is so much duplication between file-based and console-based and I didn't really understand the reasoning from the linked PR.

**Comment by carltongibson** [comment]
Hey @jschneier. Thanks for the follow-up. 

I didn't get a chance to look at this again yet. I will do so in the next week. 

I can't see much of a BC issue here... (?!)

**Comment by jschneier** [comment]
> I can't see much of a BC issue here... (?!)

In the same way that the tests depends on what is output (and thus are currently failing) couldn't someone's code be depending on it?

**Comment by jschneier** [comment]
> Regarding c988745, it looks to me like there aren't any tests for that change. I rolled back to that commit and reverted the changes in django.core.mail and all tests pass on Python 2 and 3.

Maybe we can ask @apollo13?

**Comment by carltongibson** [comment]
> ...couldn't someone's code be depending on it?

Yes. Hmmm. I suppose so. But then I steer back towards thinking this should just be a wontfix, plus “create your own backend if BCCs are important to you”. (I kind of agree with you that the hook idea is suboptimal — we should just fix it — but if BC is a concern then how else do we satisfy both requirements?)

**Comment by jschneier** [comment]
I was thinking a release note message with the code to restore.

**Comment by apollo13** [comment]
@jschneier https://github.com/django/django/pull/2126 The main reason was consistency. I had an issue with my emails and wanted to debug it with the console backend. This did not work because the unicode representation used a different encoding then when writing to bytes. Imo my change still makes sense, even if there are not tests showing it.

**Comment by jschneier** [comment]
Yeah that’s fine. More unsure about if the change to file-based is needed. Changing the mode to `a` from `ab` would mean we don’t need to override `write_message` and since we aren’t specifying an encoding I assume that it ends up being written out in non-`b` mode the same? Namely with `sys.getfilesystemencoding()`. Not my area of expertise but would make sense to me logically, please let me know if I’ve missed something.

**Comment by apollo13** [comment]
@jschneier It should be fine to write to the file in non-binary mode as long as you use `as_bytes` and then `decode` (ie just don't use `as_string`; which will result in a different serialization then for a normal smtp transaction). Maybe also add a comment on why the bytes -> decode dance is there that it doesn't get replaced by accident.

**Comment by jschneier** [comment]
Just opened #10241  pursuant to this discussion. Probably if we really want to mimic the smtp backend we should also add `linesep='\r\n'`? Appreciate any feedback there.

**Comment by apollo13** [comment]
Good catch, yes. I think I added `\r\n` after I rewrote the filebackend and forgot to add it there :)

**Comment by jschneier** [comment]
Interestingly using `linesep='\r\n'` results in test failures on Windows.

**Comment by apollo13** [comment]
Mind showing which? I don't have a windows system currently :)

**Comment by jschneier** [comment]
Me neither.

I pushed up the changes in the aforementioned PR including `linesep=‘\r\n’` and it failed 15 tests on the file-based backend which I think are all of the ones that depend on `get_mailbox_content` on the Windows sqlite build config. I’ve since force-pushed over it.

**Comment by carltongibson** [comment]
Right, lets see if we can roll this up... (It shouldn't be _that_ complicated... 😄) 

@jschneier: What do you **want** to do here? What's the idea outcome for you? 

@all: can we accept a BC change? Surely it's easier for people to test the logic that generates the arguments to the `send_mail()` call (e.g. `recipients_list`) rather than call re-parsing the output later (?)) (If so the change suggested here looks fine to me in principle.) 

Failing that: My idea for the `format_message()` hook was easier subclassing with a behaviour change. If that's not worth the effort (which it may not be) then I'm inclined towards `wontfix`. I had thought even the doc change wasn't really necessary but, since Josh ran into it to, it probably is worth mentioning. 

If we nail down the strategy, I can help debug the windows issues etc. (A bit of de-duplication here seems feasible.)

**Comment by felixxm** [comment]
Closing due to inactivity and lack of consensus.

**Pull request closed without merging**